dev/benchmarks/macrobenchmarks/test/util.dart

```diff
@@ -0,0 +1,220 @@
+        find.byKey(const ValueKey<String>(kScrollableName));
+    expect(scrollable, findsOneWidget);
+    final Finder button =
+        find.byKey(ValueKey<String>(routeName), skipOffstage: false);
+    await tester.ensureVisible(button);
+    expect(button, findsOneWidget);
+    await tester.pumpAndSettle();
+    await tester.tap(button);
+
+    if (pageDelay != null) {
+      // Wait for the page to load
+      await tester.binding.delayed(pageDelay);
+    }
+
+    if (setup != null) {
+      await setup(tester);
+    }
+
+    await watchPerformance(binding, () async {
+      final Future<void> durationFuture = tester.binding.delayed(duration);
+      if (body != null) {
+        await body(tester);
+      }
+      await durationFuture;
+    });
+  }, semanticsEnabled: false, timeout: Timeout(timeout));
+}
+
+Future<void> watchPerformance(
+  E2EWidgetsFlutterBinding binding,
+  Future<void> action(),
+) async {
+  final List<FrameTiming> frameTimings = <FrameTiming>[];
+  final TimingsCallback watcher = frameTimings.addAll;
+  binding.addTimingsCallback(watcher);
+  await action();
+  binding.removeTimingsCallback(watcher);
+  // TODO(CareF): determine if it's running on firebase and report metric online
+  final FrameTimingSummarizer frameTimes = FrameTimingSummarizer(frameTimings);
+  binding.reportData = <String, dynamic>{'performance': frameTimes.summary};
+}
+
+/// This class and summarizes a list of [FrameTiming] for the performance
+/// metrics.
+class FrameTimingSummarizer {
+  factory FrameTimingSummarizer(List<FrameTiming> data) {
+    assert(data != null);
+    assert(data.isNotEmpty);
+    final List<Duration> frameBuildTime = List<Duration>.unmodifiable(
+      data.map<Duration>((FrameTiming datum) => datum.buildDuration),
+    );
+    final List<Duration> frameBuildTimeSorted = List<Duration>.from(frameBuildTime)..sort();
+    final List<Duration> frameRasterizerTime = List<Duration>.unmodifiable(
+      data.map<Duration>((FrameTiming datum) => datum.rasterDuration),
+    );
+    final List<Duration> frameRasterizerTimeSorted = List<Duration>.from(frameBuildTime)..sort();
+    final Duration Function(Duration, Duration) add = (Duration a, Duration b) => a + b;
+    return FrameTimingSummarizer._(
+      frameBuildTime,
+      frameRasterizerTime,
+      // This avarage calculation is microsecond precision, which is fine
+      // because typical values of these times are milliseconds.
+      frameBuildTime.reduce(add) ~/ data.length,
+      _findPercentile(frameBuildTimeSorted, 0.90),
+      _findPercentile(frameBuildTimeSorted, 0.99),
+      frameBuildTimeSorted.last,
+      _countExceed(frameBuildTimeSorted, kBuildBudget),
+      frameRasterizerTime.reduce(add) ~/ data.length,
+      _findPercentile(frameRasterizerTimeSorted, 0.90),
+      _findPercentile(frameRasterizerTimeSorted, 0.90),
+      frameRasterizerTimeSorted.last,
+      _countExceed(frameRasterizerTimeSorted, kBuildBudget),
+    );
+  }
+
```

Comment:
Nit: when a function has a lot of arguments, Flutter usually put them under `{}` so it would be like the following
```
const FrameTimingSummarizer._({
  this.frameBuildTime,
  ...
});

// Use it with argument name provided to reduce confusion
FrameTimingSummarizer._(
  ...,
  averageFrameBuildTime: frameBuildTime.reduce(add) ~/ data.length,
  ...,
);
```

Reply:
`{}` means named arguments, and optional if not marked as `@required`, with default value if not assigned `null`. Here I don't think they should be optional or named since this private constructor is just above its call and should only be used here.